kclvm/compiler_base/error/src/lib.rs

```diff
@@ -1,5 +1,488 @@
+//! ```
+
+mod diagnostic;
+mod diagnostic_handler;
 mod emitter;
 
-pub use emitter::Emitter;
-pub use emitter::TerminalEmitter;
+use anyhow::{bail, Context, Result};
+use diagnostic_handler::DiagnosticHandlerInner;
+use fluent::FluentArgs;
+use std::sync::Mutex;
+
+pub use diagnostic::{components, style::DiagnosticStyle, Component, Diagnostic};
+pub use emitter::{Emitter, TerminalEmitter};
+
+#[cfg(test)]
+mod tests;
+
+/// `DiagnosticHandler` supports diagnostic messages to terminal stderr.
+///
+/// `DiagnosticHandler` will load template file directory when instantiating through the constructor `new_with_template_dir()`.
+///
+/// Note: `DiagnosticHandler` uses `Mutex` internally to ensure thread safety,
+/// so you don't need to use references like `Arc` or `Mutex` to make `DiagnosticHandler` thread safe.
+///
+/// When your compiler needs to use `Compiler-Base-Error` to displaying diagnostics, you need to create a `DiagnosticHandler` at first.
+/// For more information about how to create a `DiagnosticHandler`, see the doc above method `new_with_template_dir()`.
+/// Since creating `DiagnosticHandler` needs to load the locally template (*.ftl) file, it may cause I/O performance loss,
+/// so we recommend you create `DiagnosticHandler` globally in the compiler and pass references to other modules that use `DiagnosticHandler`.
+///
+/// And since `DiagnosticHandler` provides methods that do not supports mutable references "&mut self", so passing immutable references (&) is enough.
+///
+/// For Example:
+///
+/// 1. You can put `DiagnosticHandler` on the same level as `Lexer`, `Parser` and `CodeGenerator` in your compiler.
+/// ```ignore
+/// struct Compiler {
+///     diag_handler: DiagnosticHandler,
+///     lang_lexer: Lexer,
+///     lang_parser: Parser,
+///     code_generator: CodeGenerator
+/// }
+/// ```
+///
+/// 2. And send the immutable references to `Lexer`, `Parser` and `CodeGenerator` to displaying the diagnostic during compiling.
+/// ```ignore
+/// impl Compiler {
+///     fn compile(&self) {
+///         self.lang_lexer.lex(&self.diag_handler);
+///         self.lang_parser.parse(&self.diag_handler);
+///         self.code_generator.gen(&self.diag_handler);
+///     }
+/// }
+/// ```
+///
+/// ```ignore
+/// impl Lexer {
+///     fn lex(&self, diag_handler: &DiagnosticHandler){
+///        handler.XXXX(); // do something to diaplay diagnostic.
+///     }
+/// }
+/// ```
+///
+pub struct DiagnosticHandler {
+    handler_inner: Mutex<DiagnosticHandlerInner>,
+}
+
+impl DiagnosticHandler {
+    /// Load all (*.ftl) template files under default directory.
+    ///
+    /// Default directory "./src/diagnostic/locales/en-US/"
+    /// Call the constructor 'new_with_template_dir()' to load the file.
+    /// For more information about the constructor 'new_with_template_dir()', see the doc above 'new_with_template_dir()'.
+    pub fn default() -> Result<Self> {
```

Comment:
U should use `impl Default for DiagnosticHandler`

Reply:
pub fn default() -> Result<Self> 
`default()` in `trait Default` need returning `Self`.